### ISSUE TITLE
poor performance with autosizing of columns if there are merged cells

### ISSUE THREAD

**Issue by sbraaa** [PHPOffice/PhpSpreadsheet#2871]
I found out that autosizing all columns in presence of merged cell really slow down xlsx creation.
I made some try using the same script (7 columns and 700 rows) :
 - merging first 4 rows in 1 column -> 215 seconds
 - without any merge : 4 seconds

anyone faced the same behaviour?
any suggests?
thanks

**Comment by Pierstoval** [comment]
I'm encountering very very strange behaviors in my current PhpSpreadsheet usage. In the past, it was a bit slow but still fine with my use-cases, but now it's becoming like extra-slow.

Like this case for instance:
| code | cli output |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/3369266/187917089-ca9d9503-c8d4-46d0-b9cd-c88444225e32.png) | ![image](https://user-images.githubusercontent.com/3369266/187917175-d2e08f32-9e88-4046-a8b6-c1d44e0cb25f.png) |

I mean, 68 seconds just to insert 13 rows? That's insane. If I insert these rows one by one instead of using the `$numberOfRows` 2nd argument of the `insertNewRowBefore()` method, it's way worse, and I have no idea why.

**Comment by MarkBaker** [comment]
Why are you asking about slow performance when inserting rows in a question about autosizing columns?

> I mean, 68 seconds just to insert 13 rows? That's insane. If I insert these rows one by one instead of using the $numberOfRows 2nd argument of the insertNewRowBefore() method, it's way worse, and I have no idea why.

When you insert (or delete) rows or columns, PhpSpreadsheet needs to search through the content of every cell in the worksheet, to see if there are any formulae that need updating; as well as various other updates like row styling, page breaks, etc. The more cells you have, the longer this will take. If you do this 13 times because you're inserting each row one at a time, then it will take 13 times longer than doing it for a single insert of 13 rows.

It sounds like you have a lot more rows/cells than you think you do. Just because a cell looks empty when you look at it in MS Excel, doesn't mean that it is empty. Cells might contain a null value, or an empty string, or even a formula that evaluates to null or an empty string; or may have conditional content where the condition isn't being met.
Are you loading a template file to update and save? If so, check the template. MS Excel has a feature to remove trailing "empty" rows/columns

**Comment by sbraaa** [comment]
> Why are you asking about slow performance when inserting rows in a question about autosizing columns?
> 
> > I mean, 68 seconds just to insert 13 rows? That's insane. If I insert these rows one by one instead of using the $numberOfRows 2nd argument of the insertNewRowBefore() method, it's way worse, and I have no idea why.
> 
> When you insert (or delete) rows or columns, PhpSpreadsheet needs to search through the content of every cell in the worksheet, to see if there are any formulae that need updating; as well as various other updates like row styling, page breaks, etc. The more cells you have, the longer this will take. If you do this 13 times because you're inserting each row one at a time, then it will take 13 times longer than doing it for a single insert of 13 rows.
> 
> It sounds like you have a lot more rows/cells than you think you do. Just because a cell looks empty when you look at it in MS Excel, doesn't mean that it is empty. Cells might contain a null value, or an empty string, or even a formula that evaluates to null or an empty string; or may have conditional content where the condition isn't being met. Are you loading a template file to update and save? If so, check the template. MS Excel has a feature to remove trailing "empty" rows/columns

Actually the question by @Pierstoval is is not relevant if we talk about auto-sizing.
@MarkBaker gives the right answer but incidentally that answer is related the issue I opened.
Looking at code it seems that slow performances when using auto sizing suffers from the same problem. To dimension correctly a column all the rows/columns must be examined. the more rows we have more time is required. 
My customers using the procedure which raise my issue not so often but it seems strange that they never notices the slowness  before (and hardware hasn't changed).
I hope someone could optimize the code, maybe better, I hope to find the time to do it myself.

**Comment by Pierstoval** [comment]
> Are you loading a template file to update and save? If so, check the template. MS Excel has a feature to remove trailing "empty" rows/columns

This was my issue! I emptied all cells "around" the editing area where my template is supposed to have data, and it went blazing fast afterwards!

**Comment by rtek** [comment]
We're experiencing significant performance issues with the cell merging after moving from 1.24 to 1.25. Approximately a 10 increase in spreadsheet generation that requires multiple merges of blank cells.

An xdebug profile traced it to this:

https://github.com/PHPOffice/PhpSpreadsheet/compare/1.24.1...1.25.0#diff-50511008e409893e5826291c61bad6a05b641e4a81dfb56da7c9b5b4f97a2ed5R1819

Any recommendations for how to bypass this extra processing for cells that are not defined or already empty?

Which merge option is most similar to the previous behavior?

**Comment by MarkBaker** [comment]
> Edit: it appears that `Worksheet::MERGE_CELL_CONTENT_HIDE` is most similar the previous behavior which is no longer the default?

Incorrect: For MS Excel, the behaviour is always to remove the values from all but the upper left cell, which is the default behaviour that I implemented (`Worksheet::MERGE_CELL_CONTENT_EMPTY`). The latest versions of Excel even tell you this when you do a merge:
![image](https://user-images.githubusercontent.com/770298/202706231-efadf566-e00b-4171-abe6-a982574266ef.png)
The previous behaviour of PhpSpreadsheet was the same as MS Excel, a Cell Merge removed the values from all but that one top-left cell.

Open/Libre Office allow the other options when merging cells; and if that file is then loaded into MS Excel, and the cells UnMerged, the previous cell values will be present.


Looking at the code, I can see some optimisations that I can make: In particular for line 1825, it's only necessary to get the formatted value if the mode is `Worksheet::MERGE_CELL_CONTENT_MERGE`, otherwise it's enough just to get the value`.
I'll take a look at this sometime over the weekend or next week.

**Comment by rtek** [comment]
Ah understood.  Is the performance issue due to using the calculated value instead of the raw value for populating the left most cell?

**Comment by MarkBaker** [comment]
It calculates the value if that cell contains a formula, but also formats the result as a string. That's unnecessary overhead for the default, it only matters if the mode is `Worksheet::MERGE_CELL_CONTENT_MERGE`: in 999 cases out of a 1000, formatting won't make any difference, but in that 0.1% of cases it would also be wrong. So I'll certainly correct that error, and that will also help performance.

There's also a redundant `if` test in the logic, a test that will always result in true; so that can be eliminated.
